src/google/protobuf/generated_message_tctable_gen.cc

```diff
@@ -251,170 +99,33 @@ bool IsFieldEligibleForFastParsing(
     return false;
   }
 
-  // We will check for a valid auxiliary index range later. However, we might
-  // want to change the value we check for inlined string fields.
-  int aux_idx = entry.aux_idx;
-
   switch (field->type()) {
       // Some bytes fields can be handled on fast path.
     case FieldDescriptor::TYPE_STRING:
     case FieldDescriptor::TYPE_BYTES:
-      if (field->options().ctype() == FieldOptions::STRING) {
-        // strings are fine...
-      } else if (field->options().ctype() == FieldOptions::CORD) {
-        // Cords are worth putting into the fast table, if they're not repeated
-        if (field->is_repeated()) return false;
-      } else {
+      if (field->options().ctype() != FieldOptions::STRING) {
         return false;
       }
-      if (options.is_string_inlined) {
-        ABSL_CHECK(!field->is_repeated());
-        // For inlined strings, the donation state index is stored in the
-        // `aux_idx` field of the fast parsing info. We need to check the range
-        // of that value instead of the auxiliary index.
-        aux_idx = entry.inlined_string_idx;
-      }
       break;
-
-    case FieldDescriptor::TYPE_ENUM: {
-      uint8_t rmax_value;
-      if (!message_options.uses_codegen &&
-          GetEnumRangeInfo(field, rmax_value) == EnumRangeInfo::kNone) {
-        // We can't use fast parsing for these entries because we can't specify
-        // the validator.
-        // TODO: Implement a fast parser for these enums.
-        return false;
-      }
+    case FieldDescriptor::TYPE_ENUM:
+      if (!cpp::HasPreservingUnknownEnumSemantics(field)) return false;
```

Comment:
Why did you make this change?

Reply:
The fast loop path of enums / varints is the same for proto3. For proto2 we have to do additional stuff so it needs a fallback